src/mots/module.py

```diff
@@ -0,0 +1,356 @@
+        )
+        if machine_readable_nick in nicks or not machine_readable_nick:
+            continue
+        nicks.append(machine_readable_nick)
+        try:
+            p.yaml_set_anchor(machine_readable_nick)
+        except Exception as e:
+            # NOTE: this happens when the file changes and we lose ruamel embedded data
+            # TODO: fix this
+            logger.error(e)
+
+    if write:
+        file_config.write()
+
+
+def validate(config: dict, repo_path: str):
+    """Validate the current state of the config file.
+
+    - Check if top-level dictionary contains required keys
+    - Check if machine names are unique
+    - Instantiate and run validation on each :class:`Module` instance
+
+    :raises ValidationError: if any validation errors are detected
+    """
+    # Validate that config has all the required keys.
+    required_keys = ("repo", "created_at", "updated_at", "modules")
+    keys_diff = required_keys - config.keys()
+    if len(keys_diff) != 0:
+        raise ValidationError(f"{keys_diff} missing from configuration file.")
+
+    modules = config["modules"]
+
+    # Count machine name repetitions in a defaultdict.
+    machine_names = defaultdict(int)
+    for m in modules:
+        machine_names[m["machine_name"]] += 1
+        if "submodules" in m and m["submodules"]:
+            for sm in m["submodules"]:
+                machine_names[sm["machine_name"]] += 1
+
+    machine_names = {name: count for name, count in machine_names.items() if count > 1}
+    if machine_names:
+        raise ValidationError(
+            f"Duplicate machine name(s) found: {', '.join(machine_names.keys())}"
+        )
+
+    errors = []
+    for i in range(len(config["modules"])):
+        module = config["modules"][i]
+        module = Module(repo_path=repo_path, **module)
+        validation_errors = module.validate()
+        if validation_errors:
+            errors.append(validation_errors)
+
+    if errors:
+        raise ValidationError(errors)
+    logger.info("All modules validated successfully.")
+
+
+def add(
+    new_module: dict, file_config: FileConfig, parent: str = None, write: bool = True
+):
+    """Add a new module to the configuration.
+
+    :param module: a dictionary containing module parameters
+    :param file_config: an instance of :class:`FileConfig`
+    :param parent: the machine name of the parent module if applicable
+    :param write: if set to `True`, writes changes to disk
+    """
+    file_config.load()
+    modules = file_config.config["modules"]
+    serialized = Module(**new_module, repo_path=file_config.repo_path).serialize()
+
+    if parent:
+        for module in modules:
+            if module["machine_name"] == parent:
+                if "submodules" not in module or not module["submodules"]:
+                    module["submodules"] = []
+                module["submodules"].append(serialized)
+                break
+    else:
+        modules.append(serialized)
```

Comment:
Is there any validation we can/should do to provided data?

Reply:
Yes, we could trigger `clean` or `validate` after this. I think I will rewrite this as part of improving overall general and schema validation

Reply:
Filed [a bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1753716) for this.